### PR TITLE
optimize upsample performance linear mode on CPU

### PR DESCRIPTION
**Pull request by mingfeima**
This pr aims at improving  `nn.UpSample()` performance on CPU with mode `linear`, `bilinear`, `trilinear`.
For single socket inference, up to **31x** performance improvement.
For single core inference, up to **1.8x** performance improvement.
For dual socket training, up to **28x** performance improvement.

`channel last` format kernel also provided.

### CODE REVIEW

**Comment by mingfeima** [comment]
Follow up on #31452

`"cubic"` mode optimization and `channel last` memory format optimization (if necessary) will be handled in other PRs.

Benchmark is available from [op_bench-py](https://github.com/mingfeima/op_bench-py). The data is collected on Xeon Skylake 6148, 2*20 cores@2.4GHz.

To reproduce, for inference data: `./run.sh upsample_linear.py`; for training data: `./run.sh upsample_linear.py --train`

**Unit**: ms per each iteration, the lower the better.

#### single socket inference

operation | input size | original | this pr | speedup
-- | -- | -- | -- | --
upsample_nearest1d | [64, 9, 512] | 3.768 | 0.138 | 27.30
upsample_nearest2d | [8, 9, 64, 64] | 7.606 | 0.411 | 18.51
upsample_nearest3d | [8, 9, 32, 32, 32] | 291.662 | 9.355 | 31.18

#### single core inference

operation | input size | original | this pr | speedup
-- | -- | -- | -- | --
upsample_nearest1d | [64, 9, 512] | 3.688 | 2.354 | 1.57
upsample_nearest2d | [8, 9, 64, 64] | 7.631 | 6.548 | 1.17
upsample_nearest3d | [8, 9, 32, 32, 32] | 275.978 | 153.596 | 1.80

#### dual core training

operation | input size | original | this pr | speedup
-- | -- | -- | -- | --
upsample_nearest1d | [64, 9, 512] | 8.031 | 0.604 | 13.30
upsample_nearest2d | [8, 9, 64, 64] | 19.273 | 1.152 | 16.73
upsample_nearest3d | [8, 9, 32, 32, 32] | 639.977 | 22.860 | 28.00

**Comment by yf225** [comment]
@VitalyFedyunin @ngimel Would you like to review this PR? Thanks!

**Comment by mingfeima** [comment]
#### update on nearest mode with channel last format
1. adjust grain size, the old grain size equals to `at::internal::GRAIN_SIZE * channels` which will prevent efficient usage of multiple cpu cores.
2. explicit vectorization

Test platform is `Xeon 8280 CPU 2*20 cores @ 2.70GHz`, benchmark [upsample_nearest.py](https://github.com/mingfeima/op_bench-py/blob/master/upsample_nearest.py)

Single socket receives **2.9x** speedup, single core receives **1.2x** speedup

```bash
(before, single socket)
upsample_nearest2d_cl: memory format:   nhwc, input size:  torch.Size([8, 9,   64, 64])
input.is_contiguous(memory_format=torch.channels_last):  True
input.is_contiguous():  False
forward time per iteration:   0.287 ms

(after, single socket)
upsample_nearest2d_cl:   memory format: nhwc, input size:    torch.Size([8, 9, 64, 64])
input.is_contiguous(memory_format=torch.channels_last):  True
input.is_contiguous():  False
forward time per iteration:   0.098 ms

(before, single core)
upsample_nearest2d_cl:   memory format: nhwc, input size:    torch.Size([8, 9, 64, 64])
input.is_contiguous(memory_format=torch.channels_last):  True
input.is_contiguous():  False
forward time per iteration:   1.770 ms

(after, single core)
upsample_nearest2d_cl:   memory format: nhwc, input size:    torch.Size([8, 9, 64, 64])
input.is_contiguous(memory_format=torch.channels_last):  True
input.is_contiguous():  False
forward time per iteration:   1.385 ms
```

**Comment by mingfeima** [comment]
#### update on linear mode with channel last format
Here also list performance of `nchw` to compare against `nhwc`. Actually `nhwc` should be treated as a turbo mode since it is more suitable for vectorization.

**comments:** 1. enlarge channel size to fully show the power of channel last format in the benchmark, vectorization is done on channel dim. 2. the old implementation has non-contiguous memory access pattern for nchw format, so the corresponding speedup ratio is even bigger than number of cores.

```bash
(before single socket)
upsample_bilinear:   memory format: nchw, input size:  [32,   128, 64, 64]
input.is_contiguous(memory_format=torch.channels_last):  False
input.is_contiguous():  True
forward time per iteration:   2260.420 ms

upsample_bilinear_cl: memory format:   nhwc, input size:  [32, 64, 64, 128]
input.is_contiguous(memory_format=torch.channels_last):  True
input.is_contiguous():  False
forward time per iteration:   2186.436 ms

(after single socket)
upsample_bilinear:   memory format: nchw, input size:  [32,   128, 64, 64]
input.is_contiguous(memory_format=torch.channels_last):  False
input.is_contiguous():  True
forward time per iteration:   19.432 ms

upsample_bilinear_cl: memory format:   nhwc, input size:  [32, 64, 64, 128]
input.is_contiguous(memory_format=torch.channels_last):  True
input.is_contiguous():  False
forward time per iteration:   5.859 ms

(before, single core)
upsample_bilinear:   memory format: nchw, input size:  [32,   128, 64, 64]
input.is_contiguous(memory_format=torch.channels_last):  False
input.is_contiguous():  True
forward time per iteration:   2321.760 ms

upsample_bilinear_cl: memory format:   nhwc, input size:  [32, 64, 64, 128]
input.is_contiguous(memory_format=torch.channels_last):  True
input.is_contiguous():  False
forward time per iteration:   2403.946 ms

(after, single core)
upsample_bilinear:   memory format: nchw, input size:  [32,   128, 64, 64]
input.is_contiguous(memory_format=torch.channels_last):  False
input.is_contiguous():  True
forward time per iteration:   365.771 ms

upsample_bilinear_cl: memory format:   nhwc, input size:  [32, 64, 64, 128]
input.is_contiguous(memory_format=torch.channels_last):  True
input.is_contiguous():  False
forward time per iteration:   72.455 ms

**Comment by mingfeima** [comment]
#### Updates
1. add `Channels Last` format kernel for `linear` mode, parallel on `num_batches` dim and vectorize on `channel` dim.
2. adjust grain size of `Channels Last` format kernel for `nearest` mode to `at::internal::GRAIN_SIZE / channels`, add explicit vectorization.

Test platform, Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz, 2 * 20 cores. To reproduce, use [upsample_linear.py](https://github.com/mingfeima/op_bench-py/blob/master/upsample_linear.py)

#### upsample linear mode (unit: ms)
config | memory format | before | after | speedup
-- | -- | -- | -- | --
single socket (omp_num_threads=20) | NCHW | 2260.42 | 19.43 | 116.3
single socket (omp_num_threads=20) | NHWC | 2186.44 | 5.86 | 373.2
single core (omp_num_threads=1) | NCHW | 2321.76 | 365.77 | 6.3
single core (omp_num_threads=1) | NHWC | 2403.95 | 72.46 | 33.2

Input tensor size is enlarged to `[32, 128, 64, 64]` to fully show the performance diff between `nhwc` and `nchw` layout, (`nhwc` layout is easier for vectorization).

**Comment by mingfeima** [comment]
@VitalyFedyunin @ngimel could you please review this PR?

File: aten/src/ATen/native/cpu/UpSampleKernel.cpp [hunk 1]
```diff
@@ -165,67 +165,64 @@ void cpu_upsample_nearest_channels_last(
...
+          ih * input_width * channels + iw * channels;
+      copy(output_ptr, input_ptr, channels);
       data_index_step(n, num_batches, od, output_depth, oh, output_height, ow, output_width);
     }
   };
 
   if (ndim == 4) {
     // upsample nearest 2d
-    at::parallel_for(0, num_work_items, at::internal::GRAIN_SIZE, loop2d);
+    at::parallel_for(0, numel / channels, at::internal::GRAIN_SIZE / channels, loop2d);
```
Comment (ngimel) [inline aten/src/ATen/native/cpu/UpSampleKernel.cpp:221]: Do shape check make sure that channels is not 0? Otherwise the code will crash here

File: aten/src/ATen/native/cpu/UpSampleKernel.cpp [hunk 2]
```diff
@@ -165,67 +165,64 @@ void cpu_upsample_nearest_channels_last(
...
-      auto input_data_ptr = input_data
-        + n * (input_depth * input_height * input_width * channels)
-        + id * (input_height * input_width * channels)
-        + ih * (input_width * channels) + iw * channels;
-      auto output_data_ptr = output_data
-        + n * (output_depth * output_height * output_width * channels)
-        + od * (output_height * output_width * channels)
-        + oh * (output_width * channels) + ow * channels;
-      std::memcpy(output_data_ptr, input_data_ptr, sizeof(scalar_t) * channels);
+      scalar_t* output_ptr = output_data + i * channels;
```
Comment (ngimel) [inline aten/src/ATen/native/cpu/UpSampleKernel.cpp:210]: cool simplification

File: aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp [hunk 3]
```diff
@@ -0,0 +1,575 @@
...
+    auto input_indexr = [=](int64_t c, int64_t w) {
+      return input_data[c * input_width + w];
+    };
+
+    int64_t iw0, iw1;
+    scalar_t w0lambda, w1lambda;
+    for (int64_t c = begin; c < end; c++) {
+      for (int64_t ow = 0; ow < output_width; ow++) {
+        compute_source_index_and_lambda(
+            iw0, iw1, w0lambda, w1lambda, width_scale, ow, input_width, output_width, align_corners);
```
Comment (ngimel) [inline aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp:82]: Do you guys care about bfloat16 here? Computing w?lambda coefficients in low precision may not be enough, but it can be left as a followup.

File: aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp [hunk 4]
```diff
@@ -0,0 +1,575 @@
...
+  }
+}
+
+using scale_t = std::vector<c10::optional<double>>;
+void upsample_linear1d_kernel_impl(
+    Tensor& output,
+    const Tensor& input,
+    bool align_corners,
+    c10::optional<double> scales_w) {
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(input.scalar_type(), "upsample_linear1d", [&] {
```
Comment (ngimel) [inline aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp:492]: Why are you dispatching half type but not bfloat16? I though half arithmetic is too slow on the cpu

File: aten/src/ATen/native/cpu/UpSampleKernel.cpp (hunk 1 above)
Comment (mingfeima) [reply aten/src/ATen/native/cpu/UpSampleKernel.cpp:222]: add check for `channels`.

File: aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp (hunk 3 above)
Comment (mingfeima) [reply aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp:82]: Yes, we do care about bf16. Let us leave bf16 as a follow up to align with bigger scope.

File: aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp (hunk 4 above)
Comment (mingfeima) [reply aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp]: Indeed it doesn't make much sense to dispatch to **half**. It's a purely copy/paste from the old code [here](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/UpSampleLinear1d.cpp#L196).
Anyway I have all **half** dispatch in upsample_xxx, let me know if you have a second thought.

**Comment by mingfeima** [comment]
@ngimel please review the updates!

File: aten/src/ATen/native/cpu/UpSampleKernel.cpp [hunk 5]
```diff
@@ -216,6 +216,7 @@ void cpu_upsample_nearest_channels_last(
     }
   };
 
+  TORCH_INTERNAL_ASSERT(channels > 0);
```
Comment (ngimel) [inline aten/src/ATen/native/cpu/UpSampleKernel.cpp:219]: This should be TORCH_CHECK, because it is triggered by user providing invalid data. `TORCH_INTERNAL_ASSERT` is for behavior when there is an internal error in pytorch, the message there is "File a bug for pytorch", this is not what should happen here.

File: aten/src/ATen/native/cpu/UpSampleKernel.cpp [hunk 6]
```diff
@@ -325,7 +326,7 @@ void upsample_nearest1d_kernel_impl(
     Tensor& output,
     const Tensor& input,
     c10::optional<double> scales_w) {
-  AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::Byte, input.scalar_type(), "upsample_nearest1d", [&] {
+  AT_DISPATCH_FLOATING_TYPES_AND(at::ScalarType::Byte, input.scalar_type(), "upsample_nearest1d", [&] {
```
Comment (ngimel) [inline aten/src/ATen/native/cpu/UpSampleKernel.cpp:330]: Out of curiosity, "byte" here is for some sort of quantization support?

File: aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp [hunk 7]
```diff
@@ -332,6 +332,7 @@ void cpu_upsample_linear_channels_last(
     }
   };
 
+  TORCH_INTERNAL_ASSERT(output_slice_size > 0);
```
Comment (ngimel) [inline aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp:336]: Same here, it's not internal pytorch error, it's user providing invalid data.

File: aten/src/ATen/native/cpu/UpSampleKernel.cpp (hunk 5 above)
Comment (mingfeima) [reply aten/src/ATen/native/cpu/UpSampleKernel.cpp]: Thanks for the heads up! Fixed.

File: aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp (hunk 7 above)
Comment (mingfeima) [reply aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp]: `upsample_{123}d_shape_check` from [UpSample.h](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/UpSample.h#L59) will check if `width`, `height` and `depth`  are greater than 0, this should guarantee that `output_slice_size`  here is greater than 0. Thus we only need to make sure `channels > 0` for the **channels_last** kernels.

File: aten/src/ATen/native/cpu/UpSampleKernel.cpp (hunk 6 above)
Comment (mingfeima) [reply aten/src/ATen/native/cpu/UpSampleKernel.cpp:330]: I think the intention is to make `interpolate` be able to accept **uint8** dtype to process raw image, not from the aspect of quantization, same story from [2289](https://github.com/pytorch/vision/issues/2289).
By the way, I didn't add the support for uint8. My original commit only covers floating points. And the request is from a medical equipment company using UNet to process their CT scan images.

**Review by ngimel — APPROVED** [review]
